vllm/distributed/communication_op.py

```diff
@@ -6,9 +6,33 @@
 from .parallel_state import get_tp_group
 
 
-def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+@torch.library.custom_op("vllm::tp_out_of_place_ar",
+                         mutates_args=["input_"],
+                         device_types=("cuda", "cpu"))
+def _tp_out_of_place_ar(input_: torch.Tensor) -> torch.Tensor:
     """All-reduce the input tensor across model parallel group."""
-    return get_tp_group().all_reduce(input_)
+    return get_tp_group().out_of_place_ar(input_)
+
+
+@torch.library.register_fake("vllm::tp_out_of_place_ar")
+def _tp_out_of_place_ar_fake(input_: torch.Tensor) -> torch.Tensor:
+    return input_
+
+
+@torch.library.custom_op("vllm::tp_in_place_ar",
+                         mutates_args=["input_"],
+                         device_types=("cuda", "cpu"))
+def _tp_in_place_ar(input_: torch.Tensor) -> None:
+    """All-reduce the input tensor across model parallel group."""
+    get_tp_group().in_place_ar(input_)
+
+
+def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+    if get_tp_group().should_run_out_of_place_ar(input_):
```

Comment:
how does Dynamo work with this condition?

Reply:
I tested this by running the `pytest tests/distributed/test_basic_distributed_correctness.py` test with dynamo enabled. Everything runs to completion without any errors when fullgraph is True or False.

Is there something specific you are worried about? I'm also happy to run additional tests.

Reply:
can you try to follow https://github.com/vllm-project/vllm/blob/baa5467547a758af35f442af6edfbc0fb73c83ce/tests/tpu/test_compilation.py#L13 and see what's the transformed bytecode?

Reply:
okay, I think all these conditions turn into guards.